lib/awful/layout/dynamic/base.lua

```diff
@@ -0,0 +1,322 @@
+    local added, removed = {}, {}
+
+    -- Get all clients visible on the tag screen, this include clients
+    -- from other selected tags
+    local clients, reverse = client.tiled(tag.getscreen(self._tag)), {}
+
+    for k,c in ipairs(clients) do
+        if check_tiled(c) and not self.client_to_wrapper[c] then
+            table.insert(added, c)
+        end
+        reverse[c] = true
+    end
+
+    for c,_ in pairs(self.client_to_wrapper) do
+        if not reverse[c] then
+            table.insert(removed, c)
+        end
+    end
+
+    return added, removed
+end
+
+-- When a tag is selected or the layout change for this one, activate the handler
+local function wake_up(self)
+    if (not self) or not self.is_dynamic then return end
+
+    if self.widget.wake_up then
+        self.widget:wake_up()
+    end
+
+    local added, removed = get_client_differential(self)
+
+    for k, c in ipairs(added) do
+        if check_tiled(c) then
+            if self.client_to_wrapper[c] then
+                self.client_to_wrapper[c]:wake_up()
+                self.widget:add(self.client_to_wrapper[c])
+            else
+                local wrapper = l_wrapper(c)
+                wrapper._handler = self
+
+                insert_wrapper(self, c, wrapper)
+            end
+        end
+    end
+
+    for k, c in ipairs(removed) do
+        local wrapper = self.client_to_wrapper[c]
+
+        remove_wrapper(self, c, wrapper)
+    end
+
+    self.active = true
+end
+
+-- When a tag is hidden or the layout isn't the handler, stop all processing
+local function suspend(self)
+    if not self.is_dynamic then return end
+
+    if self.widget.suspend then
+        self.widget.suspend(self.widget)
+    end
+
+    self.active = false
+end
+
+-- Emulate the main "layout" method of a hierarchy
+local function main_layout(self, handler)
+
+    if not handler.param then
+        handler.param = aw_layout.parameters(handler._tag)
+        handler.param.is_init = true
+    end
+
+    local workarea = handler.param.workarea
+
+    handler.hierarchy:update(get_context(tag.getscreen(handler._tag)), handler.widget, workarea.width, workarea.height)
+
+end
+
+local function nop() end
+local context_index_miss = setmetatable({
+    operator            = 0,
+    rectangle           = nop,
+    clip                = nop,
+    reset_clip          = nop,
+    push_group          = nop,
+    pop_group           = nop,
+    pop                 = nop,
+    pop_group_to_source = nop,
+    paint_with_alpha    = nop,
+    get_matrix          = function(self) return self                end,
+    clip_extents        = function(self) return {width=0, height=0} end,
+    transform           = function(self, cmatrix)
+        return self:multiply(matrix.from_cairo_matrix(cmatrix))
+    end,
+}, {__index = matrix})
+
+-- Emulate enough of cairo context to use unmodied hierarchies
+local function create_fake_context()
+    local m = matrix.identity:copy()
+    m:set_mutating(true)
```

Comment:
...to make a matrix look like a cairo context. Sounds a lot like a hack. Also, the cairo API is a lot larger than this (random examples: `:set_operator()`, `.target`).

Oh and `:clip_extents()` is supposed to return two numbers, not a table. And when you fix this (and keep returning `0,0`), this whole code will break since `wibox.hierarchy:draw` sees "the clip is empty, so I don't have to draw anything" (clip extents are the basis on how partial redraws are implemented).

All together: NO!